File: prev_exams_and_ans/2015ccc/q4_wait_time/WaitTime.java
import java.io.IOException;
//import java.util.Scanner;
//import java.util.HashMap;
//import java.util.ArrayList;
import java.util.*;

public class WaitTime {
	
	// <friendId, [waitTime, indicator]}>
	// indicator: -1 if all cleared up; timeStamp if not cleared up;
	private static HashMap <String, int[]> waitTimeTable = new HashMap<>();
	private static ArrayList<String> friendArray = new ArrayList<>();
	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
		int rows = Integer.parseInt(sc.nextLine());
		int timeStamp = 0;
		for (int i = 0; i < rows; i++) {
			String[] rowInfo = sc.nextLine().split(" ");
			String action = rowInfo[0];

			if (!action.equals("W")) {
				// action equals R or S
				String friendId = rowInfo[1];
				if (!waitTimeTable.containsKey(friendId) /* new friend */) {
					int[] friendInfo = {0, timeStamp};
					waitTimeTable.put(friendId, friendInfo);
					friendArray.add(friendId);
				} else /* old friend */ {
					// old friend, either "S" to complete conversation 
					//                 or "R" to start a new conversation
					int[] friendInfo = waitTimeTable.get(friendId);
					if (action.equals("S")) {
						friendInfo[0] = friendInfo[0] + (timeStamp - friendInfo[1]);
						friendInfo[1] = -1; // indicator: -1; complete conversation.
						waitTimeTable.put(friendId, friendInfo);
					} else /* "R" */ {
						friendInfo[1] = timeStamp;
						waitTimeTable.put(friendId, friendInfo);
					}
				}
				timeStamp += 1;
			} else /* action equals W*/ {
				timeStamp += Integer.parseInt(rowInfo[1]) - 1; // Why -1?
			} 
			
		}
		
		Collections.sort(friendArray);
		for (String friendId: friendArray) {
			int[] friendInfo = waitTimeTable.get(friendId);
			System.out.println(friendId + " " + (friendInfo[1] == -1 ? friendInfo[0] : -1));
		}
	}

}

File: prev_exams_and_ans/2015ccc/q4_wait_time/wait_time.py
# <friendId, [waitTime, indicator(-1 or timeStamp)]}>
# #  indicator: -1 if all cleared up; timeStamp if not cleared up;

wait_time_table = {}

rows = int(input())
time_stamp = 0
for i in range(0, rows):
    rowInfo = input().split()
    action = rowInfo[0]

    if (action != "W"):
        # action equals R or S
        friendId = rowInfo[1]
        if friendId not in wait_time_table:
            friendInfo = [0, time_stamp]
            wait_time_table[friendId] = friendInfo
        else: # old friend
            # old friend, either "S" to complete conversation
            #                  or "R" to start a new conversation
            friendInfo = wait_time_table[friendId]
            if (action == "S"):
                friendInfo[0] = friendInfo[0] + (time_stamp - friendInfo[1])
                friendInfo[1] = -1 # indicator: -1; complete conversation
                wait_time_table[friendId] = friendInfo
            else: # /"R"/ 
                friendInfo[1] = time_stamp
                wait_time_table[friendId] = friendInfo
        time_stamp += 1
    else: # action equals W
        time_stamp += int(rowInfo[1]) - 1

friend_array = list(wait_time_table.keys())
friend_array.sort()
for friend in friend_array:
    friendInfo = wait_time_table[friend]
    print(friend, friendInfo[0] if friendInfo[1] == -1 else -1)